### ISSUE TITLE
Rewriting doesn't check preconditions

### ISSUE THREAD

**Issue by IanOrton** [agda/agda#2059]
It seems that rewrite rules don't check any preconditions. Is this expected behaviour?

``` agda
{-# OPTIONS --rewriting #-}

module contradiction where

open import Data.Bool
open import Data.Empty
open import Relation.Binary.Core
open import Relation.Nullary.Negation

obvious : (b b' : Bool) (p : b ≢ b') → (b ∧ b') ≡ false
obvious false b'    notEq = refl
obvious true  false notEq = refl
obvious true  true  notEq = contradiction refl notEq

{-# BUILTIN REWRITE _≡_ #-} 
{-# REWRITE obvious   #-}

oops : (b : Bool) → b ∧ b ≡ false
oops b = refl

true≢false : true ≡ false → ⊥
true≢false ()

bot : ⊥
bot = true≢false (oops true)
```

This typechecks on Agda 2.5.1

**Comment by andreasabel** [comment]
In principle, rewriting is unsafe, there are no solid checks of preconditions.

What is a bit funny for this example that `obvious` is rejected if I replace one of the `refl`s in its definition by a hole, e.g.

``` agda
obvious false b'    notEq = {! refl !}
```

fails with

```
obvious  is not a legal rewrite rule, 
since the following variables are not bound by the left hand side:  p
when checking the pragma REWRITE obvious
```

@jespercockx Did not you recently make sure all bound variables are used on the lhs of the rewrite rule, even the non-dependent `p` ?

**Comment by jespercockx** [comment]
Yes, I added the check that all variables used by the rewrite rule are bound in the LHS. However, somehow in this example the third argument of `obvious` gets marked as an `UnusedArg`. 

I think I misunderstood the meaning of `UnusedArg` in Agda: I thought it meant that it is not matched on or used in the rhs, but it rather means that the function is _constant_ in this argument, a much weaker property. For example Agda considers the argument of type `⊥` in `⊥-elim` to be unused.

```
⊥-elim : ∀ {w} {Whatever : Set w} → ⊥ → Whatever
⊥-elim ()
```

In other words, we cannot reuse the relevance tags for checking whether a rewrite rule uses a particular argument. Do you know if Agda internally has another way to check whether an argument is used?

**Comment by andreasabel** [comment]
In internal syntax, an absurd pattern just becomes a variable pattern. Thus the positivity checker cannot see a match going on here.  This, of course, can be changed, to take absurd matches into account.  For a conservative change, we could refine the `Relevance` type to parametrize the `Unused` by a bit whether it was unused in the current sense, or whether it is unused but for absurd matches.

**Comment by andreasabel** [comment]
I am afraid I will not be able to work on this until Monday (even one week after).  However, I think the relevance/polarity stuff needs some clarification.  I think we have the following aspects muddled together:
1. Irrelevant args: things irrelevant at compile (and thus also at runtime). Explicitly declared by the user.  Can be skipped during definitional equality checking.
2. Forced args: typing prevents only one possible value. Relevant at runtime (and thus at compile-time). Inferred. Can be skipped during equality check.
3. Nonstrict args: Cannot move to evaluation position. Irrelevant at runtime but relevant at compile time. Declared. Cannot be skipped during equality check.  Could be the correct notion of "type argument", akin to Conor's multiplicity 0.
4. Unused args: Relevant at compile time but irrelevant at runtime. Inferred. Skipped during equality check. Absurd matching allowed (like for 1. irrelevant args).

I feel we should sort out things according to these dimensions:
1. Declared vs. inferred.
2. Relevant at runtime.
3. Relevant at compile time.  (Not independent of 2.)
4. Skipped during equality check.
5. Droppable (meaning this argument can be dropped completely at compile time without endangering typing or evaluation).  Necessarily includes 2.-4., thus, maybe not a dimension.

Further, polarities should be included in this picture...

**Comment by jespercockx** [comment]
I'm trying to fix this by making absurd matches not count as a UnusedArg (after all, they should be desugarable to an applications of ⊥-elim, so the argument is not actually unused). That fixes this issue, however it causes test/Succeed/NonvariantPolarity.agda to fail. In particular,

```
module Absurd where

  -- Absurd patterns do not count as matches; abort is constant in its 2nd arg.
  abort : (A : Set) → ⊥ → A
  abort A ()

  test : (x y : ⊥) → abort Bool x ≡ abort Bool y
  test x y = refl
```

now causes the following error:

```
x != y of type ⊥
when checking that the expression refl has type
abort Bool x ≡ abort Bool y
```

I'm not sure if this kind of 'irrelevance' in arguments of the empty type is something we want, at least I'm quite sure that it's not normally part of the rules of MLTT. So I would be tempted to just remove this from the testsuite and see if anyone complains. Objections?

**Comment by andreasabel** [comment]
I am fine with removing this test (or put it into Fail).
